vm/compiler/src/ledger/helpers/mod.rs

```diff
@@ -0,0 +1,401 @@
+///     H_Y1 = Anchor block height at year 1.
+pub const fn staking_reward<const STARTING_SUPPLY: u64, const ANCHOR_TIME: u16>() -> u64 {
+    // Compute the anchor block height at year 1.
+    let anchor_height_at_year_1 = anchor_block_height(ANCHOR_TIME, 1);
+    // Compute the annual staking reward: (0.025 * S).
+    let annual_staking_reward = (STARTING_SUPPLY / 1000) * 25;
+    // Compute the staking reward: (0.025 * S) / H_Y1.
+    annual_staking_reward / anchor_height_at_year_1 as u64
+}
+
+/// Calculates the coinbase reward for a given block.
+///     R_coinbase = max(0, H_Y10 - H) * R_anchor * 2^(-1 * (D - B) / B).
+///     R_anchor = Anchor reward.
+///     H_Y10 = Anchor block height at year 10.
+///     H = Current block height.
+///     D = Time elapsed since the previous block.
+///     B = Anchor block time.
+pub fn coinbase_reward<const STARTING_SUPPLY: u64, const ANCHOR_TIME: u16, const NUM_BLOCKS_PER_EPOCH: u32>(
+    previous_timestamp: i64,
+    timestamp: i64,
+    block_height: u32,
+) -> Result<u64> {
+    // Compute the anchor block height at year 10.
+    let anchor_height_at_year_10 = anchor_block_height(ANCHOR_TIME, 10);
+    // Compute the anchor reward.
+    let anchor_reward = anchor_reward::<STARTING_SUPPLY, ANCHOR_TIME>();
+    // Compute the remaining blocks until year 10, as a u64.
+    let num_remaining_blocks_to_year_10 = anchor_height_at_year_10.saturating_sub(block_height) as u64;
+    // Return the coinbase reward.
+    match num_remaining_blocks_to_year_10.checked_mul(anchor_reward).ok_or_else(|| anyhow!("Anchor reward overflow"))? {
+        // After the anchor block height at year 10, the coinbase reward is 0.
+        0 => Ok(0),
+        // Until the anchor block height at year 10, the coinbase reward is determined by this equation:
+        //   (num_remaining_blocks_to_year_10 * anchor_reward) * 2^{-1 * ((timestamp - previous_timestamp) - ANCHOR_TIME) / ANCHOR_TIME}
+        reward => retarget::<ANCHOR_TIME>(reward, previous_timestamp, timestamp, ANCHOR_TIME as u32, true),
+    }
+}
+
+/// Calculates the anchor reward.
+///     R_anchor = floor((2 * S) / (H_Y10 * (H_Y10 + 1))).
+///     S = Starting supply.
+///     H_Y10 = Expected block height at year 10.
+const fn anchor_reward<const STARTING_SUPPLY: u64, const ANCHOR_TIME: u16>() -> u64 {
+    // Calculate the anchor block height at year 10.
+    let anchor_height_at_year_1 = anchor_block_height(ANCHOR_TIME, 10) as u64;
+    // Compute the numerator.
+    let numerator = 2 * STARTING_SUPPLY;
+    // Compute the denominator.
+    let denominator = anchor_height_at_year_1 * (anchor_height_at_year_1 + 1);
+    // Return the anchor reward.
+    numerator / denominator
+}
+
+/// Returns the anchor block height after a given number of years for a specific anchor time.
+pub const fn anchor_block_height(anchor_time: u16, num_years: u32) -> u32 {
+    // Calculate the number of seconds in a year.
+    const SECONDS_IN_A_YEAR: u32 = 60 * 60 * 24 * 365;
```

Comment:
(I'm assuming that leap years can be ignored)

Reply:
This logic is used for the first decade (approximately), thus it should witness at most 2 leap days, which is acceptable. (For context, the retarget algorithm is a fixed-point approximation with an error bound greater than 2 / (365 * 10) * 100 %)